File: LeetCode/101-200/113. 路径总和 II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    List<Integer> list = new ArrayList<>();
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        dfs(root, 0, sum);
        return res;
    }
    public void dfs(TreeNode root, int num, int sum){
        if(root == null) return;
        num += root.val;
        list.add(root.val);
        if(num == sum && root.left == null && root.right == null) res.add(new ArrayList(list));
        dfs(root.left, num, sum);
        dfs(root.right, num, sum);
        list.remove(list.size() - 1);
    }
}

File: LeetCode/113. 路径总和 II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function (root, targetSum) {
  if (!root) return []
  const result = []
  const queue = [[root, targetSum, [root.val]]]
  while (queue.length) {
    const n = queue.length
    for (let i = 0; i < n; i++) {
      const [r, target, arr] = queue.shift()
      const nextTarget = target - r.val
      if (!r.left && !r.right && nextTarget === 0) {
        result.push(arr)
      } else {
        if (r.left) queue.push([r.left, nextTarget, [...arr, r.left.val]])
        if (r.right) queue.push([r.right, nextTarget, [...arr, r.right.val]])
      }
    }
  }
  return result
}
